client-spark/shuffle-manager-2/src/main/java/org/apache/spark/shuffle/rss/RssShuffleManager.java

```diff
@@ -0,0 +1,221 @@
+      "org.apache.spark.shuffle.sort.SortShuffleManager";
+
+  private final SparkConf conf;
+  private final RssConf rssConf;
+  private final int cores;
+  private String newAppId;
+
+  private LifecycleManager lifecycleManager;
+  private ShuffleClient rssShuffleClient;
+  private volatile SortShuffleManager _sortShuffleManager;
+  private final ConcurrentHashMap.KeySetView<Integer, Boolean> sortShuffleIds =
+      ConcurrentHashMap.newKeySet();
+  private final RssShuffleFallbackPolicyRunner fallbackPolicyRunner;
+
+  public RssShuffleManager(SparkConf conf) {
+    this.conf = conf;
+    this.rssConf = SparkUtils.fromSparkConf(conf);
+    this.cores = conf.getInt(SparkLauncher.EXECUTOR_CORES, 1);
+    this.fallbackPolicyRunner = new RssShuffleFallbackPolicyRunner(conf);
+  }
+
+  private boolean isDriver() {
+    return "driver".equals(SparkEnv.get().executorId());
+  }
+
+  private SortShuffleManager sortShuffleManager() {
+    if (_sortShuffleManager == null) {
+      synchronized (this) {
+        if (_sortShuffleManager == null) {
+          _sortShuffleManager =
+              SparkUtils.instantiateClass(sortShuffleManagerName, conf, isDriver());
+        }
+      }
+    }
+    return _sortShuffleManager;
+  }
+
+  private void initializeLifecycleManager(String appId) {
+    // Only create LifecycleManager singleton in Driver. When register shuffle multiple times, we
+    // need to ensure that LifecycleManager will only be created once. Parallelism needs to be
+    // considered in this place, because if there is one RDD that depends on multiple RDDs
+    // at the same time, it may bring parallel `register shuffle`, such as Join in Sql.
+    if (isDriver() && lifecycleManager == null) {
+      synchronized (this) {
+        if (lifecycleManager == null) {
+          lifecycleManager = new LifecycleManager(appId, rssConf);
+          rssShuffleClient =
+              ShuffleClient.get(
+                  lifecycleManager.self(), rssConf, lifecycleManager.getUserIdentifier());
+        }
+      }
+    }
+  }
+
+  @Override
+  public <K, V, C> ShuffleHandle registerShuffle(
+      int shuffleId, int numMaps, ShuffleDependency<K, V, C> dependency) {
+    // Note: generate newAppId at driver side, make sure dependency.rdd.context
+    // is the same SparkContext among different shuffleIds.
+    // This method may be called many times.
+    newAppId = SparkUtils.genNewAppId(dependency.rdd().context());
+    initializeLifecycleManager(newAppId);
+
+    if (fallbackPolicyRunner.applyAllFallbackPolicy(
+        lifecycleManager, dependency.partitioner().numPartitions())) {
+      logger.warn("Fallback to SortShuffleManager!");
+      sortShuffleIds.add(shuffleId);
+      return sortShuffleManager().registerShuffle(shuffleId, numMaps, dependency);
+    } else {
+      // If not driver, use dummy rss meta service host and port.
+      String metaServiceHost = "";
+      int metaServicePort = 0;
+      if (lifecycleManager != null) {
```

Comment:
lifecycleManager will not be null here since initializeLifecycleManager is called before

Reply:
> lifecycleManager will not be null here since initializeLifecycleManager is called before

Change both in 2 and 3.